### ISSUE TITLE
`PwBaseWorkChain`: sanity check on bands will always fail the restart for `nscf` or `bands` calculations

### ISSUE THREAD

**Issue by sphuber** [aiidateam/aiida-quantumespresso#489]
The reason is that the sanity check will force the restart to be from scratch, because if you change the number of bands, you cannot reuse the existing charge density or wavefunctions. However, starting from scratch in a bands or nscf calculation also does not make any sense and despite specifying `from_scratch` the code will attempt to read the XML to restart, which is not there and so the code crashes.

**Comment by sphuber** [comment]
Not sure what the correct approach should be here. The problem is that if one decides restarting with more bands is the correct option, then one has to perform an `scf` calculation first. Since this is a different calculation then the `PwBaseWorkChain` was tasked with, it cannot do it itself, but this should be delegated up the call stack.

I see a few options on how to handle insignificant bands during an nscf calculation, all with  downsides:

1. Simply skip the band occupation sanity check for a `nscf` or `bands` calculation.
2. Change the sanity handler to only emit a warning but then not do anything
3. Let the sanity check actually fail the work chain, but with a specific exit code. The calling process can then check for this exit code and decide to rerun an `scf` with a higher number of bands before retrying to run the `bands` calculation.

@giovannipizzi @greschd love your input on what you think makes most sense

**Comment by greschd** [comment]
I'm not sure I completely understand the issue: Why isn't it possible to increase the number of bands in the bands calculation only? Isn't that exactly what is done in the `PwBandsWorkChain` with the `nbands_factor` logic?

Other than that, option 3 seems the most sensible to me (for the case of `nscf` and `bands`, increasing the bands in the `scf` case).
If the preceding `scf` has the same sanity check, the `nscf` or `bands` typically shouldn't even run into this error, right?

**Comment by sphuber** [comment]
> If the preceding scf has the same sanity check, the nscf or bands typically shouldn't even run into this error, right?

That is true, only if the increased number of bands in the `scf` base workchain are carried over to the `nscf` one. However, this will depend on the caller of the `PwBaseWorkChain`. This was already implemented for the `PwRelaxWorkChain` but the `PwBandsWorkChain` did not properly propagate the updated number of bands. As you pointed out, this is exactly what is addressed by PR #490 . However, the underlying problem will persist if someone just directly runs a `PwBaseWorkChain` in `nscf` or `bands` mode. So that is why I open this issue as I don't think the work chain should crash as it currently does in that case.

The difficulty then is that the real fix, running a new scf with increased number of bands, is outside the scope of the `PwBaseWorkChain`. Because it would require a lot of additional logic: essentially running a completely new `scf` calculation and _if_ that worked, resume with the originally requested `nscf` calculation. However, the new `scf` calculation should also just be run in a `PwBaseWorkChain` to make sure errors are handled, because it can fail just as well. This is why I think, the real fix should be handled upstream the call-stack and the best way of doing this is, I think so far, is to communicate this problem with a special exit code.

**Comment by greschd** [comment]
Yes, I agree with that. The special exit code seems to be the most sensible choice - it communicates clearly that there was a serious issue, but if the calling code decides they want to use the results anyway they are still free to do so.

**Comment by sphuber** [comment]
Actually, my original analysis of the problem is incorrect. I assumed that restarting an `nscf` with more bands than of the calculation from which we are restarting. I got this idea because this is the behavior that the sanity check is currently enforcing. By setting the `parent_folder` to `None` a `from_scratch` calculation will be done. However, I think I added this functionality by mistake when I first added the sanity check. It is perfectly possible to change the number of bands in a restart, also for non-scf calculations. I probably must have confused it with other parameters, that when changed invalidate the restart files, such as the cutoffs. In that case the wavefunctions and grids are no longer commensurate and cannot be reused. The fix should then just be to always restart normally when increasing the bands in the sanity check
